openmicroscopy/openmicroscopy
# New role webadmin

# What this PR does

Webadmin UI for setting Admin Privileges.

Key Changes:
 - 'Role' field to choose between 'User', 'Administrator' or 'Restricted Admin'
 - 'Role' field is disabled if you are editing yourself (as before: can't remove yourself from being Admin)
 - Can't remove 'System' group from the groups chooser (no X to click). This would select 'User' role and also cause the 'restricted admin' privilege list to disappear which would be a bit disconcerting.
 - When saving as 'Administrator', ALL Privileges will get set to 'true'.
 - When saving as 'User' role, privileges are not wiped clean on server BUT they appear to be all unset when you edit the user again, since ```adminService.getAdminPrivileges()``` returns empty list.
 - 'Write' privilege includes 'WriteOwned', 'WriteFile' & 'WriteManagedRepo'. IF just 1 or 2 of these are set to 'true' via the API, the UI will show 'Write' as unchecked. Save will result in all 3 privileges -> false. Same for 'Delete', and 'Upload Script' (Write/DeleteScriptRepo).

# Testing this PR

 - Create a user with 'Role': 'User', 'Administrator' or 'Restricted Admin'.
    - See that 'system' group is added to 'Groups' chooser when 'Admin' or 'Restricted Admin' chosen.
 - Choose from list of Privileges.
 - Save then 'Edit' user again to see what's been saved. Check 'Role' and 'Privileges' as expected.
 - Edit user with different combinations - again, check what's Saved is preserved.
 - Try setting/getting Admin Privileges via API and checking in UI, especially with partial 'Write' / 'Delete' privileges as described above.

# Related reading

See: https://github.com/openmicroscopy/design/issues/62

![screen shot 2017-02-15 at 14 12 04](https://cloud.githubusercontent.com/assets/900055/23018043/7c91d7fe-f434-11e6-9d53-a84004e2c9cd.png)


 TODO:
 - [x] hide the privileges list to only show it alongside 'Role' when 'Restricted Admin' is chosen.
 - [x] Make sure we save changes in correct order so that we don't have Admin with restrictions failed to save.
 - [x] Fix help url to point to docs page (when we know the url)!
